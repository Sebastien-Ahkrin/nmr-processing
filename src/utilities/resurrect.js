export function resurrect(acsString) {
  const data = {
    experiment: {},
    ranges: [],
    acsString,
    normalized: acsString
      .replace(/[\r\n\t]/g, ' ')
      .replace(/[;:]/g, ',')
      .replace(/\}/g, ')')
      .replace(/\{/g, '(')
      .replace(/[\u2011\u2012\u2013\u2014\u2015\u2212]/g, '-'),
  };

  createParts(data);
  parseParts(data);
  return data;
}

function parseParts(data) {
  for (const part of data.parts) {
    if (part.toLowerCase().includes('nmr')) {
      processExperiment(data, part);
    } else {
      processRange(data, part);
    }
  }
}

function processRange(data, part) {
  const split = splitParenthesis(part);
  // before parenthesis there should be only numbers but we will still split with space
  const beforeParts = split.before
    .split(/(?: |(?<=[0-9])-)/)
    .map((part) => Number(part))
    .filter((part) => !Number.isNaN(part));
  if (beforeParts.length < 1 || beforeParts.length > 2) return;
  const from = beforeParts[0];
  const to = beforeParts.length > 1 ? beforeParts[1] : beforeParts[0];

  const insideParts = split.inside.split(/ *, */);
  let signal = { js: [] };
  let range = { from, to, signals: [signal] };

  const integrationParts = insideParts.filter((part) =>
    part.match(/^[0-9]+H$/),
  );
  if (integrationParts.length === 1) {
    range.integration = Number(integrationParts[0].replace('H', ''));
  }

  const multiplicityParts = insideParts.filter((part) =>
    part.match(/^[a-zA-Z]+$/),
  );
  if (multiplicityParts.length === 1) {
    const multiplicity = multiplicityParts[0];
    if (multiplicity === 'm') {
      if (beforeParts.length === 1) {
        // a complex signal
        signal.delta = beforeParts[0];
        signal.multiplicity = multiplicity;
      } else {
        // a real range
        signal.delta = (beforeParts[0] + beforeParts[1]) / 2;
        signal.multiplicity = multiplicity;
      }
    } else {
      // looks like a real multiplicity, s, d, dd, etc..
      if (beforeParts.length === 1) {
        // a complex signal
        signal.delta = beforeParts[0];
        signal.multiplicity = multiplicity;
      }
    }
  }

  const jCouplings = insideParts
    .filter((part) => part.match(/(Hz|J|^[0-9.]+$)/))
    .map((jCoupling) => Number(jCoupling.replace(/[^0-9.]/g, '')));

  const multiplicities = (signal.multiplicity || '').split(/(?=.)/); // will have to be improved for quint, hex, hept, ...
  if (multiplicities.length === jCouplings.length) {
    for (let i = 0; i < multiplicities.length; i++) {
      signal.js.push({
        coupling: jCouplings[i],
        multiplicity: multiplicities[i],
      });
    }
  }

  data.ranges.push(range);
}

function processExperiment(data, part) {
  const split = splitParenthesis(part);
  const before = split.before
    .replace(/[ -]*nmr[ -]*/i, '')
    .replace(/[ -]/g, '');
  if (before.match(/^[0-9]+[A-Z][a-z]?$/)) {
    // 36Cl, 1H, 13C, ...
    data.experiment.nucleus = before;
  }
  if (before.match(/^[A-Z][a-z]?[0-9]+$/)) {
    // Cl35, H1, C13, ...
    data.experiment.nucleus = before.replace(/^([A-Z][a-z]?)([0-9]+)$/, '$2$1');
  }
  if (split.inside) {
    // some frequency and solvent ???
    const insideParts = split.inside.split(/[,]/);
    const frequencyParts = insideParts.filter((part) => part.match(/[0-9]{2}/));
    if (frequencyParts.length) {
      const frequency = frequencyParts[0].replace(/[^[0-9.]]/g, '');
      if (frequency.length > 1) data.experiment.frequency = Number(frequency);
    }
    const solventParts = insideParts.filter((part) => !part.match(/[0-9]{2}/));
    if (solventParts.length) {
      data.experiment.solvent = solventParts[0];
    }
  }
}

function createParts(data) {
  const parts = data.normalized
    .split(/\)(?![^()]*\))/)
    .map((part) => part.replace(/^\s*(.*?)\s*$/, '$1'))
    .filter((part) => part)
    .map((part) => `${part})`)
    .map((part) => part.split(/[,;](?![^()]*\))/))
    .flat()
    .map((part) => part.replace(/^\s*(.*?)\s*$/, '$1'))
    .filter((part) => part);

  data.parts = parts;
}

function splitParenthesis(part) {
  if (!part.includes('(')) return { before: part, inside: '', after: '' };
  if (!part.includes(')')) {
    throw new Error(`A part does not contain closing parenthesis: ${part}`);
  }
  return part.match(
    / *(?<before>[^(]*?) *\( *(?<inside>.*?) *\) *(?<after>.*?) */,
  ).groups;
}
