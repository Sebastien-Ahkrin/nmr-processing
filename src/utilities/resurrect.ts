import type { Range } from '../types/range';
import { resurrectRange } from './resurrectRange';

export interface DataResurrect {
  experiment: any;
  ranges: Range[];
  acsString: string;
  normalized: string;
  parts: string[];
}

export function resurrect(acsString: string) {
  const data: any = {
    experiment: {},
    ranges: [],
    acsString,
    normalized: acsString
      .replace(/[\r\n\t]/g, ' ')
      .replace(/[;:]/g, ',')
      .replace(/\}/g, ')')
      .replace(/\{/g, '(')
      .replace(/[\u2011\u2012\u2013\u2014\u2015\u2212]/g, '-'),
  };
  createParts(data);
  parseParts(data);
  return data;
}

function parseParts(data: DataResurrect) {
  for (const part of data.parts) {
    if (part.toLowerCase().includes('nmr')) {
      processExperiment(data, part);
    } else {
      const range = resurrectRange(part);
      if (range) data.ranges.push(range);
    }
  }
}

function processExperiment(data: DataResurrect, part: any) {
  const split = splitParenthesis(part);
  const before = split.before
    .replace(/[ -]*nmr[ -]*/i, '')
    .replace(/[ -]/g, '');
  if (before.match(/^[0-9]+[A-Z][a-z]?$/)) {
    // 36Cl, 1H, 13C, ...
    data.experiment.nucleus = before;
  }
  if (before.match(/^[A-Z][a-z]?[0-9]+$/)) {
    // Cl35, H1, C13, ...
    data.experiment.nucleus = before.replace(/^([A-Z][a-z]?)([0-9]+)$/, '$2$1');
  }
  if (split.inside) {
    // some frequency and solvent ???
    const insideParts = split.inside.split(/[,]/);
    const frequencyParts = insideParts.filter((part: any) =>
      part.match(/[0-9]{2}/),
    );
    if (frequencyParts.length) {
      const frequency = frequencyParts[0].replace(/[^[0-9.]]/g, '');
      if (frequency.length > 1) data.experiment.frequency = Number(frequency);
    }
    const solventParts = insideParts.filter(
      (part: any) => !part.match(/[0-9]{2}/),
    );
    if (solventParts.length) {
      data.experiment.solvent = solventParts[0];
    }
  }
}

function createParts(data: DataResurrect) {
  const parts = data.normalized
    .split(/\)(?![^()]*\))/)
    .map((part) => part.replace(/^\s*(.*?)\s*$/, '$1'))
    .filter((part) => part)
    .map((part) => `${part})`)
    .map((part) => part.split(/[,;](?![^()]*\))/))
    .flat()
    .map((part) => part.replace(/^\s*(.*?)\s*$/, '$1'))
    .filter((part) => part);
  data.parts = parts;
}
