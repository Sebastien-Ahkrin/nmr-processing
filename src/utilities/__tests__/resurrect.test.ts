/* eslint-disable no-tabs */

import { resurrect } from '../resurrect';

describe('resurrect', () => {
  it.skip('angewandte Chemie', () => {
    const string = `
	1H-NMR (400 MHz, CDCl3) 5.96 (dd, J1 = 10.0 Hz, J2 = 5.4 Hz, 1 H), 5.71 (dd, J1 = 
	10.1 Hz, J2 = 1.1 Hz, 1 H), 5.14 (br. s, 1 H), 4.71 (dd, J1 = 11.1 Hz, J2 = 3.7 Hz, 1 H), 
	4.07 (sept., J = 6.2 Hz, 1 H), 2.41 (dd, J1 = 17.7 Hz, J2 = 11.2 Hz, 1 H), 2.23 (dd, J1 = 
	17.7 Hz, J2 = 4.1 Hz, 1 H), 1.29 (d, J = 6.2 Hz, 3 H), 1.19 (d, J = 6.1 Hz, 3 H), 1.00 (t, 
	J = 7.8 Hz, 9 H), 0.63 (q, J = 7.8 Hz, 6 H).`;
    const ranges = resurrect(string).ranges;
    expect(ranges).toMatchSnapshot();
  });

  it.skip('JOC', () => {
    const string = `
	H-NMR (CDCl, 300 MHz) 3.68 (s, 2H), 3.70-3.74 (m, 4H), 3.93 (dd, 1H, J = 9.7, 
	J2 = 3.7), 4.49 (d, 2 H, J = 4.1), 4.56-4.61 (m, 1H), 5.12 (s, 2H), 5.81 (d, 1H, J = 7.5), 
	7.24-7.37 (m, 10H).`;
    const ranges = resurrect(string).ranges;
    expect(ranges).toMatchSnapshot();
  });

  it('Tetrahedron', () => {
    const string = `
	1H NMR (CDCl3, 400 MHz) d: 5.55 (d, J = 8.1 Hz, 1H, H-10 ), 4.72 (br s, 1H),
	4.29 (m, 1H, H-50 ), 3.53 (m, 2H, H-30 , H-40 ), 3.42  
	(m, 1H, H-20 ), 3.15 (dd, J = 6.0 Hz, J = 10.0 Hz, 1H, H-3), 3.00 (td, J = 11.0 Hz,  
	J = 4.5 Hz, 1H, H-19), 1.69 (s, 1H, H-30)
`;
    const ranges = resurrect(string).ranges;
    expect(ranges).toMatchSnapshot();
  });

  it.skip('Parenthesis', () => {
    const string = `
1H NMR (400 MHz, CDCl3): d/ppm 0.89 (t, 3H, J = 7.0 Hz, CH3), 1.23-1.45 {m, 24H, CH3(CH2)12- 
}, 1.77 (qt, 2H, J = 7.4 Hz, -CH2CH2COO-), 2.58 (t, 2H, J = 7.5 Hz, -CH2COO-), 7.14 (d, 2H, J = 
8.7 Hz, Ar-H), 7.26 (d, 2H, J = 8.7 Hz, Ar-H), 7.41 (dd, 1H, J = 7.8 Hz, 4.9 Hz, Ar-H), 8.30 (d, 1H, 
J = 7.9 Hz, Ar-H), 8.51 (s, 1H, CH=N), 8.71 (dd, 1H, J = 4.7 Hz, 1.4 Hz, Ar-H), 9.02 (d, 1H, J = 
1.4 Hz, Ar-H). 
`;
    const ranges = resurrect(string).ranges;
    expect(ranges).toMatchSnapshot();
  });

  it('J. Org. Chem.', () => {
    const string = `1H NMR (CDCl3) δ 10.58 (b, 1H), 7.40 (d, 1H, J = 8.0 Hz), 6.19 (d, 1H, J = 7.6 Hz), 4.88 (s, 1H), 4.46 (m, 1H), 4.25 (m, 1H), 4.08 (m, 1H), 3.90−3.95 (m, 2H), 2.17 (s, 3H), 1.02 (s, 9H), 1.01 (s, 9H), 0.89 (s, 9H), 0.12 (s, 3H), 0.11 (s, 3H)`;
    const ranges = resurrect(string).ranges;
    expect(ranges).toMatchSnapshot();
  });
});
