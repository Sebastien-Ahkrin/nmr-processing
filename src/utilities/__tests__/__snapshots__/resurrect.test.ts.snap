// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resurrect J. Org. Chem. 1`] = `
Array [
  Object {
    "from": 10.53,
    "integration": 1,
    "signals": Array [
      Object {
        "delta": 10.58,
        "js": Array [],
        "multiplicity": "b",
      },
    ],
    "to": 10.63,
  },
  Object {
    "from": 7.340000000000001,
    "integration": 1,
    "signals": Array [
      Object {
        "delta": 7.4,
        "js": Array [
          Object {
            "coupling": 8,
            "multiplicity": "d",
          },
        ],
        "multiplicity": "d",
      },
    ],
    "to": 7.46,
  },
  Object {
    "from": 6.1305000000000005,
    "integration": 1,
    "signals": Array [
      Object {
        "delta": 6.19,
        "js": Array [
          Object {
            "coupling": 7.6,
            "multiplicity": "d",
          },
        ],
        "multiplicity": "d",
      },
    ],
    "to": 6.2495,
  },
  Object {
    "from": 4.83,
    "integration": 1,
    "signals": Array [
      Object {
        "delta": 4.88,
        "js": Array [],
        "multiplicity": "s",
      },
    ],
    "to": 4.93,
  },
  Object {
    "from": 4.41,
    "integration": 1,
    "signals": Array [
      Object {
        "delta": 4.46,
        "js": Array [],
        "multiplicity": "m",
      },
    ],
    "to": 4.51,
  },
  Object {
    "from": 4.2,
    "integration": 1,
    "signals": Array [
      Object {
        "delta": 4.25,
        "js": Array [],
        "multiplicity": "m",
      },
    ],
    "to": 4.3,
  },
  Object {
    "from": 4.03,
    "integration": 1,
    "signals": Array [
      Object {
        "delta": 4.08,
        "js": Array [],
        "multiplicity": "m",
      },
    ],
    "to": 4.13,
  },
  Object {
    "from": 3.9,
    "integration": 2,
    "signals": Array [
      Object {
        "delta": 3.925,
        "js": Array [],
        "multiplicity": "m",
      },
    ],
    "to": 3.95,
  },
  Object {
    "from": 2.12,
    "integration": 3,
    "signals": Array [
      Object {
        "delta": 2.17,
        "js": Array [],
        "multiplicity": "s",
      },
    ],
    "to": 2.2199999999999998,
  },
  Object {
    "from": 0.97,
    "integration": 9,
    "signals": Array [
      Object {
        "delta": 1.02,
        "js": Array [],
        "multiplicity": "s",
      },
    ],
    "to": 1.07,
  },
  Object {
    "from": 0.96,
    "integration": 9,
    "signals": Array [
      Object {
        "delta": 1.01,
        "js": Array [],
        "multiplicity": "s",
      },
    ],
    "to": 1.06,
  },
  Object {
    "from": 0.84,
    "integration": 9,
    "signals": Array [
      Object {
        "delta": 0.89,
        "js": Array [],
        "multiplicity": "s",
      },
    ],
    "to": 0.9400000000000001,
  },
  Object {
    "from": 0.06999999999999999,
    "integration": 3,
    "signals": Array [
      Object {
        "delta": 0.12,
        "js": Array [],
        "multiplicity": "s",
      },
    ],
    "to": 0.16999999999999998,
  },
  Object {
    "from": 0.06,
    "integration": 3,
    "signals": Array [
      Object {
        "delta": 0.11,
        "js": Array [],
        "multiplicity": "s",
      },
    ],
    "to": 0.16,
  },
]
`;

exports[`resurrect Tetrahedron 1`] = `
Array [
  Object {
    "from": 5.489875,
    "integration": 1,
    "signals": Array [
      Object {
        "delta": 5.55,
        "js": Array [
          Object {
            "coupling": 8.1,
            "multiplicity": "d",
          },
        ],
        "multiplicity": "d",
      },
    ],
    "to": 5.610125,
  },
  Object {
    "from": NaN,
    "integration": 1,
    "signals": Array [
      Object {
        "delta": NaN,
        "js": Array [],
      },
    ],
    "to": NaN,
  },
  Object {
    "from": 4.24,
    "integration": 1,
    "signals": Array [
      Object {
        "delta": 4.29,
        "js": Array [],
        "multiplicity": "m",
      },
    ],
    "to": 4.34,
  },
  Object {
    "from": 3.48,
    "integration": 2,
    "signals": Array [
      Object {
        "delta": 3.53,
        "js": Array [],
        "multiplicity": "m",
      },
    ],
    "to": 3.5799999999999996,
  },
  Object {
    "from": 3.37,
    "integration": 1,
    "signals": Array [
      Object {
        "delta": 3.42,
        "js": Array [],
        "multiplicity": "m",
      },
    ],
    "to": 3.4699999999999998,
  },
  Object {
    "from": 3.08,
    "integration": 1,
    "signals": Array [
      Object {
        "delta": 3.15,
        "js": Array [
          Object {
            "coupling": 6,
            "multiplicity": "d",
          },
          Object {
            "coupling": 10,
            "multiplicity": "d",
          },
        ],
        "multiplicity": "dd",
      },
    ],
    "to": 3.2199999999999998,
  },
  Object {
    "from": 2.930625,
    "integration": 1,
    "signals": Array [
      Object {
        "delta": 3,
        "js": Array [
          Object {
            "coupling": 11,
            "multiplicity": "t",
          },
          Object {
            "coupling": 4.5,
            "multiplicity": "d",
          },
        ],
        "multiplicity": "td",
      },
    ],
    "to": 3.069375,
  },
  Object {
    "from": 1.64,
    "integration": 1,
    "signals": Array [
      Object {
        "delta": 1.69,
        "js": Array [],
        "multiplicity": "s",
      },
    ],
    "to": 1.74,
  },
]
`;
